L1Trigger/CSCTriggerPrimitives/src/CSCComparatorDigiFitter.cc

```diff
@@ -0,0 +1,297 @@
+
+      // normalize phi values according to first one
+      if (phis_.size()>0 and gpphi>0 and phis_[0]<0 and  (gpphi-phis_[0])>M_PI)
+        phi_tmp += (gpphi-2*M_PI);
+      else if (phis_.size()>0 and gpphi<0 and phis_[0]>0 and (gpphi-phis_[0])<-M_PI)
+        phi_tmp += (gpphi+2*M_PI);
+      else
+        phi_tmp += (csc_gp.phi());
+
+      z_tmp = csc_gp.z();
+      radius_tmp += csc_gp.perp();
+    }
+
+    //in case there are more than one comparator digis in one layer
+    radius_tmp = radius_tmp/(p.second).size();
+    radius_ += radius_tmp;
+
+    zs_.push_back(z_tmp);
+    ezs_.push_back(0);
+
+    phi_tmp = phi_tmp/(p.second).size();
+    phis_.push_back(phi_tmp);
+    ephis_.push_back(cscHalfStripWidth(detId)/sqrt(12));
+  }
+}
+
+void CSCComparatorDigiFitter::fit(const CSCDetId& ch_id, const CSCCorrelatedLCTDigi& stub, const CSCComparatorDigiCollection& hCSCComparators)
+{
+  // first, match the comparator digis to the LCT
+  matchingComparatorDigisLCT(ch_id, stub, hCSCComparators);
+
+  // second, get the coordinates
+  getComparatorDigiCoordinates(ch_id, stub);
+
+  // get radius of the stub from key layer
+  const CSCDetId key_id(ch_id.endcap(), ch_id.station(), ch_id.ring(), ch_id.chamber(), CSCConstants::KEY_CLCT_LAYER);
+  const float fractional_strip = stub.getFractionalStrip();
+
+  const auto& cscChamber = cscGeometry_->chamber(ch_id);
+  const auto& layer_geo = cscChamber->layer(CSCConstants::KEY_CLCT_LAYER)->geometry();
+
+  // LCT::getKeyWG() also starts from 0
+  const float wire = layer_geo->middleWireOfGroup(stub.getKeyWG() + 1);
+  const LocalPoint& csc_intersect = layer_geo->intersectionOfStripAndWire(fractional_strip, wire);
+  const GlobalPoint& csc_gp = cscGeometry_->idToDet(key_id)->surface().toGlobal(csc_intersect);
+
+  // get radius from key layer
+  if (useKeyRadius_)
+    radius_ = radius_/phis_.size();
+  else
+    radius_ = csc_gp.perp();
+  
+  float alpha = -99., beta = 0.;
+  // do a fit to the comparator digis
+  calculateSlopeIntercept(alpha, beta);
+  if (phis_ .size() <= 2 or std::abs(alpha)>=99){
+    alpha = csc_gp.phi();
+    beta = 0.0;
+  }
+
+  float stripPhiPitch = layer_geo->stripPhiPitch();
+  if (nStripBits_)
+    stripPhiPitch = stripPhiPitch/nStripBits_;
+
+  int fit_z_layers[6];
+  int fit_phi_layers[6];
+  for (int i=0; i<6; i++){
+    fit_z_layers[i] = cscChamber->layer(i+1)->centerOfStrip(20).z();
+    fit_phi_layers[i] = normalizedPhi(alpha + beta * fit_z_layers[i]);
+    if (nStripBits_)
+      fit_phi_layers[i] = (std::floor(fit_phi_layers[i]/stripPhiPitch) + 0.5)*stripPhiPitch;
+  }
+}
+
+void CSCComparatorDigiFitter::calculateSlopeIntercept(float& alpha, float& beta)
+{
+  if (phis_.size()>=3) {
+  
+    const bool isFront(zs_.front() < zs_.back());
+    const float zmin = isFront ? zs_.front() : zs_.back();
+    const float zmax = isFront ? zs_.back() : zs_.front();
+
+    // do the fit
+    std::unique_ptr<TF1> fit1(new TF1("fit1","pol1",zmin,zmax)); 
```

Comment:
Are there fitter classes in CMSSW that provides the same functionality that I can use?